OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,816 @@
+        // submission.
+        // We expect that each `QueueSubmission` that made it into the `queue_to_submissions_`
+        // contains at least one command buffer.
+        auto submit_info_reverse_it = submission.submit_infos.rbegin();
+        while (submit_info_reverse_it != submission.submit_infos.rend()) {
+          const SubmitInfo& submit_info = submission.submit_infos.back();
+          if (submit_info.command_buffers.empty()) {
+            ++submit_info_reverse_it;
+            continue;
+          }
+          // We found our last command buffer, so lets check if its result is there:
+          const SubmittedCommandBuffer& last_command_buffer = submit_info.command_buffers.back();
+          uint32_t check_slot_index_end = last_command_buffer.command_buffer_end_slot_index;
+
+          static constexpr VkDeviceSize kResultStride = sizeof(uint64_t);
+          uint64_t test_query_result = 0;
+          VkResult query_worked = dispatch_table_->GetQueryPoolResults(device)(
+              device, query_pool, check_slot_index_end, 1, sizeof(test_query_result),
+              &test_query_result, kResultStride, VK_QUERY_RESULT_64_BIT);
+
+          // Only erase the submission if we query its timers now.
+          if (query_worked == VK_SUCCESS) {
+            erase_submission = true;
+            completed_submissions.push_back(submission);
+          } else {
+            erase_submission = false;
+          }
+          break;
+        }
+
+        if (erase_submission) {
+          submission_it = queue_submissions.erase(submission_it);
+        } else {
+          ++submission_it;
+        }
+      }
+    }
+
+    return completed_submissions;
+  }
+
+  uint64_t QueryGpuTimestampNs(VkDevice device, VkQueryPool query_pool, uint32_t slot_index,
+                               float timestamp_period) {
+    static constexpr VkDeviceSize kResultStride = sizeof(uint64_t);
+
+    uint64_t timestamp = 0;
+    VkResult result_status = dispatch_table_->GetQueryPoolResults(device)(
+        device, query_pool, slot_index, 1, sizeof(timestamp), &timestamp, kResultStride,
+        VK_QUERY_RESULT_64_BIT);
+    CHECK(result_status == VK_SUCCESS);
+
+    return static_cast<uint64_t>(static_cast<double>(timestamp) * timestamp_period);
+  }
+
+  static void WriteMetaInfo(const SubmissionMetaInformation& meta_info,
+                            orbit_grpc_protos::GpuQueueSubmissionMetaInfo* target_proto) {
+    target_proto->set_tid(meta_info.thread_id);
+    target_proto->set_pre_submission_cpu_timestamp(meta_info.pre_submission_cpu_timestamp);
+    target_proto->set_post_submission_cpu_timestamp(meta_info.post_submission_cpu_timestamp);
+  }
+
+  void WriteCommandBufferTimings(const QueueSubmission& completed_submission,
+                                 orbit_grpc_protos::GpuQueueSubmission* submission_proto,
+                                 std::vector<uint32_t>& query_slots_to_reset, VkDevice device,
```

Comment:
`submission_proto` is passed by pointer while `query_slots_to_reset` is passed by reference? Is there a reason for doing it differently for the two? Same in `WriteDebugMarkers` below.

Reply:
Not really, that was auto suggested by the Clion refactoring. I changed, such that both are pointers.